Android aarch64 cross

compnerd opened the pull request:
<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd committed:
build: support android aarch64

This adds the needed support to enable building the standard library for android
aarch64 assuming that the appropriate parameters are passed to CMake.

compnerd committed:
build: correct the android build checks

Currently, only Darwin and Linux are supported as build hosts.  Correct the
check for the cross-compiled standard library build.  This matches the condition
used in the SDK configuration.  Move the SDK configuration step to after the
validation and the configuration of the primary variant.

compnerd committed:
build: inline android specific configuration

The android specific configuration here is the same as the non-android in the
sense that it is setting up the triples for the target.  Simply inline it as it
makes it more obvious what it is doing.  This is preparatory work for adding
support for multiple architectures for Android.

compnerd committed:
build: rename a variable to be more descriptive (NFC)

This variable is the build host, not a suffix.  Name it appropriately.  NFC.

compnerd committed:
build: remove `SWIFT_ANDROID_SDK_PATH`

The path to the SDK can be computed from the NDK.  Do so to reduce the
parameters needed to pass to CMake.  More importantly, this will allow building
with multiple architectures simultaneously.

compnerd committed:
build: hardcode the android triple in the SDK configuration

The SDK configuration can identify the triple for the android targets.  This is
similar to the Windows behaviour.  Doing so will now enable multi-architecture
builds!

drodriguez commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -207,17 +180,44 @@ macro(configure_sdk_unix
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_NDK_TRIPLE "aarch64-linux-android")
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_ALT_SPELLING "aarch64")
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_PATH "${SWIFT_ANDROID_NDK_PATH}/platforms/android-${SWIFT_ANDROID_API_LEVEL}/arch-arm64")
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_TRIPLE "aarch64-unknown-linux-android")
```
`unknown` or `none`? For consistency with ARMv7, and to not confuse people later, they should be the same (the documentation uses `none`).

compnerd replied:
Technically, the vendor should be `unknown`, not `none`.

drodriguez commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -207,17 +180,43 @@ macro(configure_sdk_unix
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_PATH "${SWIFT_ANDROID_NDK_PATH}/platforms/android-${SWIFT_ANDROID_API_LEVEL}/arch-arm64")
+        set(SWIFT_SDK_ANDROID_ARCH_${arch}_TRIPLE "aarch64-unknown-linux-android")
+      else()
+        message(FATAL_ERROR "unkonwn arch for android SDK: ${arch}")
```
Small typo in `unkonwn` (my typo, sorry).

compnerd replied:
lol, will fix :-)

compnerd committed:
build: allow control over the android arches built

Allow the user to control which of the architectures to build.  If none
is specified, then all of them will be built.

compnerd commented:
@swift-ci please test

compnerd committed:
build: update the build-script-impl to limit android to ARM

Update the build-script-impl to only build the armv7 variant by default.  This
should prevent the community buildbots from failing which may not have the arm64
ICU available.

compnerd commented:
@swift-ci please test

compnerd commented:
CC: @shahmishal @zayass @gonzalolarralde @colemancda

shahmishal commented:
@swift-ci build toolchain

compnerd commented:
@Rostepher you may care too :-)

shahmishal reviewed (approved):
LGTM! cc/ @zisko @Rostepher

shahmishal commented:
@swift-ci build toolchain
